File: ta4j-core/src/main/java/org/ta4j/core/indicators/PPOIndicator.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.indicators;

import java.util.Objects;

import org.ta4j.core.Indicator;
import org.ta4j.core.indicators.averages.EMAIndicator;
import org.ta4j.core.num.Num;

/**
 * Percentage price oscillator (PPO) indicator (also called "MACD Percentage
 * Price Oscillator (MACD-PPO)").
 *
 * @see <a href=
 *      "https://www.investopedia.com/terms/p/ppo.asp">https://www.investopedia.com/terms/p/ppo.asp</a>
 */
public class PPOIndicator extends CachedIndicator<Num> {

    private final Indicator<Num> indicator;
    private final transient EMAIndicator shortTermEma;
    private final transient EMAIndicator longTermEma;

    /**
     * Constructor with:
     *
     * <ul>
     * <li>{@code shortBarCount} = 12
     * <li>{@code longBarCount} = 26
     * </ul>
     *
     * @param indicator the indicator
     */
    public PPOIndicator(Indicator<Num> indicator) {
        this(validatedConfig(indicator, 12, 26));
    }

    /**
     * Constructor.
     *
     * @param indicator     the indicator
     * @param shortBarCount the short time frame
     * @param longBarCount  the long time frame
     */
    public PPOIndicator(Indicator<Num> indicator, int shortBarCount, int longBarCount) {
        this(validatedConfig(indicator, shortBarCount, longBarCount));
    }

    private PPOIndicator(Config config) {
        super(config.indicator());
        this.indicator = config.indicator();
        this.shortTermEma = config.shortTermEma();
        this.longTermEma = config.longTermEma();
    }

    private static Config validatedConfig(Indicator<Num> indicator, int shortBarCount, int longBarCount) {
        Indicator<Num> validatedIndicator = Objects.requireNonNull(indicator, "indicator must not be null");
        if (shortBarCount > longBarCount) {
            throw new IllegalArgumentException("Long term barCount must be greater than short term barCount");
        }
        EMAIndicator shortTermEma = new EMAIndicator(validatedIndicator, shortBarCount);
        EMAIndicator longTermEma = new EMAIndicator(validatedIndicator, longBarCount);
        return new Config(validatedIndicator, shortTermEma, longTermEma);
    }

    @Override
    protected Num calculate(int index) {
        Num shortEmaValue = shortTermEma.getValue(index);
        Num longEmaValue = longTermEma.getValue(index);
        return shortEmaValue.minus(longEmaValue)
                .dividedBy(longEmaValue)
                .multipliedBy(getBarSeries().numFactory().hundred());
    }

    @Override
    public int getCountOfUnstableBars() {
        int emaUnstableBars = Math.max(shortTermEma.getCountOfUnstableBars(), longTermEma.getCountOfUnstableBars());
        return indicator.getCountOfUnstableBars() + emaUnstableBars;
    }

    private record Config(Indicator<Num> indicator, EMAIndicator shortTermEma, EMAIndicator longTermEma) {
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/indicators/PPOIndicator.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators

import org.ta4j.core.Indicator
import org.ta4j.core.num.*

/**
 * Percentage price oscillator (PPO) indicator. <br></br>
 * Aka. MACD Percentage Price Oscillator (MACD-PPO).
 *
 * @see [https://www.investopedia.com/terms/p/ppo.asp](https://www.investopedia.com/terms/p/ppo.asp)
 */
/**
 * Constructor.
 *
 * @param indicator     the indicator
 * @param shortBarCount the short time frame
 * @param longBarCount  the long time frame
 */

open class PPOIndicator @JvmOverloads constructor(
    indicator: Indicator<Num>,
    shortBarCount: Int = 12,
    longBarCount: Int = 26
) : CachedIndicator<Num>(indicator) {

    private val shortTermEma: EMAIndicator
    private val longTermEma: EMAIndicator

    init {
        require(shortBarCount <= longBarCount) { "Long term period count must be greater than short term period count" }
        shortTermEma = EMAIndicator(indicator, shortBarCount)
        longTermEma = EMAIndicator(indicator, longBarCount)
    }

    override fun calculate(index: Int): Num {
        val shortEmaValue = shortTermEma[index]
        val longEmaValue = longTermEma[index]
        return ((shortEmaValue - longEmaValue) / longEmaValue) * numOf(100)
    }
}